src/main/java/io/cdap/plugin/gcp/dataplex/sink/DataplexOutputFormatProvider.java

```diff
@@ -0,0 +1,174 @@
+  }
+
+  /**
+   * OutputFormat for Dataplex Sink
+   */
+  public static class DataplexOutputFormat extends OutputFormat<Object, Object> {
+    private final OutputFormat<NullWritable, StructuredRecord> gcsDelegateFormat =
+      new GCSOutputFormatProvider.GCSOutputFormat();
+    private final OutputFormat<StructuredRecord, NullWritable> bqDelegateFormat = new BigQueryOutputFormat();
+    private OutputFormat delegateFormat;
+
+    /**
+     * It will set outputformat based on asset type in dataplex
+     * @param configuration
+     * @return
+     * @throws IOException
+     */
+    private OutputFormat getDelegateFormatInstance(Configuration configuration) throws IOException {
+      if (delegateFormat != null) {
+        return delegateFormat;
+      }
+      String assetType = configuration.get(DATAPLEX_ASSET_TYPE);
+      if (assetType.equalsIgnoreCase(DataplexBatchSink.BIGQUERY_DATASET_ASSET_TYPE)) {
+        delegateFormat = bqDelegateFormat;
+      } else {
+        delegateFormat = gcsDelegateFormat;
+      }
+      return delegateFormat;
+    }
+
+    @Override
+    public RecordWriter<Object, Object> getRecordWriter(TaskAttemptContext taskAttemptContext) throws
+      IOException, InterruptedException {
+      RecordWriter originalWriter = getDelegateFormatInstance(taskAttemptContext.getConfiguration())
+        .getRecordWriter(taskAttemptContext);
+      return new DataplexRecordWriter(originalWriter);
+    }
+
+    @Override
+    public void checkOutputSpecs(JobContext jobContext) throws IOException, InterruptedException {
+      getDelegateFormatInstance(jobContext.getConfiguration()).checkOutputSpecs(jobContext);
+    }
+
+    @Override
+    public OutputCommitter getOutputCommitter(TaskAttemptContext taskAttemptContext) throws IOException,
+      InterruptedException {
+      OutputCommitter delegateCommitter = getDelegateFormatInstance(taskAttemptContext.getConfiguration())
+        .getOutputCommitter(taskAttemptContext);
+      return new DataplexOutputCommitter(delegateCommitter);
+    }
+  }
+
+  /**
+   * RecordWriter for DataplexSink
+   */
+  public static class DataplexRecordWriter extends RecordWriter<Object, Object> {
+
+    private final RecordWriter originalWriter;
+    private long recordCount;
+
+    public DataplexRecordWriter(RecordWriter originalWriter) {
+      this.originalWriter = originalWriter;
+    }
+
+    @Override
+    public void write(Object keyOut, Object valOut) throws IOException,
+      InterruptedException {
+      originalWriter.write(keyOut, valOut);
+      recordCount++;
+    }
+
+    @Override
+    public void close(TaskAttemptContext taskAttemptContext) throws IOException, InterruptedException {
+      originalWriter.close(taskAttemptContext);
+      //Since the file details are not available here, pass the value on in configuration
+      taskAttemptContext.getConfiguration()
+        .setLong(String.format(RECORD_COUNT_FORMAT, taskAttemptContext.getTaskAttemptID()), recordCount);
```

Comment:
why are we keeping track of record count?

Reply:
In close method we are setting up the final recordCount in configuration.